### ISSUE TITLE
Error "Unbalanced parentheses" while relying on auto indent

### ISSUE THREAD

**Issue by pablocouto** [rust-lang/rust-mode#107]
The following code produces an `"Unbalanced parentheses"` error when pressing TAB at the blank line after `let c = "";`:

``` rust
const a: &'static str = "
{}";
fn main() {
    let b = "//";
    let c = "";

}
```

The error doesn’t occur if `const a` is defined on a single line, or either one bracket or one slash is removed from the corresponding string.

Steps I followed, with Emacs 24.5.1:
1. Start Emacs with `emacs -q`
2. Load rust-mode with `M-x load-file ...path...`
3. Load the code
4. Press TAB where indicated

Debug trace:

``` elisp
Debugger entered--Lisp error: (scan-error "Unbalanced parentheses" 77 1)
  scan-lists(77 -1 1)
  up-list(-1)
  backward-up-list()
  (save-excursion (rust-rewind-irrelevant) (backward-up-list) (rust-rewind-to-beginning-of-current-level-expr) (+ (current-column) rust-indent-offset))
  (or (if rust-indent-method-chain (progn (rust-align-to-method-chain))) (save-excursion (rust-rewind-irrelevant) (backward-up-list) (rust-rewind-to-beginning-of-current-level-expr) (+ (current-column) rust-indent-offset)))
  (if (= 0 level) 0 (or (if rust-indent-method-chain (progn (rust-align-to-method-chain))) (save-excursion (rust-rewind-irrelevant) (backward-up-list) (rust-rewind-to-beginning-of-current-level-expr) (+ (current-column) rust-indent-offset))))
  (let* ((level (rust-paren-level)) (baseline (if (= 0 level) 0 (or (if rust-indent-method-chain (progn (rust-align-to-method-chain))) (save-excursion (rust-rewind-irrelevant) (backward-up-list) (rust-rewind-to-beginning-of-current-level-expr) (+ (current-column) rust-indent-offset)))))) (cond ((nth 3 (syntax-ppss)) (let* ((string-begin-pos (nth 8 (syntax-ppss))) (end-of-prev-line-pos (if (> ... 1) (progn ...)))) (if (and (/= 114 (char-after string-begin-pos)) end-of-prev-line-pos (> end-of-prev-line-pos string-begin-pos) (= 92 (char-before end-of-prev-line-pos))) (progn (if (= ... ...) (if ... ... baseline) (save-excursion ... ... ...)))))) ((looking-at "->") (save-excursion (backward-list) (or (rust-align-to-expr-after-brace) (+ baseline rust-indent-offset)))) ((looking-at "}") (- baseline rust-indent-offset)) ((and (nth 4 (syntax-ppss)) (looking-at "*")) (+ 1 baseline)) (t (or (if (> level 0) (progn (save-excursion (rust-rewind-irrelevant) (backward-up-list) (rust-align-to-expr-after-brace)))) (progn (back-to-indentation) (if (or (looking-at "\\<else\\>\\|{") (save-excursion ... ...)) baseline (+ baseline rust-indent-offset)))))))
  (save-excursion (back-to-indentation) (let* ((level (rust-paren-level)) (baseline (if (= 0 level) 0 (or (if rust-indent-method-chain (progn ...)) (save-excursion (rust-rewind-irrelevant) (backward-up-list) (rust-rewind-to-beginning-of-current-level-expr) (+ ... rust-indent-offset)))))) (cond ((nth 3 (syntax-ppss)) (let* ((string-begin-pos (nth 8 ...)) (end-of-prev-line-pos (if ... ...))) (if (and (/= 114 ...) end-of-prev-line-pos (> end-of-prev-line-pos string-begin-pos) (= 92 ...)) (progn (if ... ... ...))))) ((looking-at "->") (save-excursion (backward-list) (or (rust-align-to-expr-after-brace) (+ baseline rust-indent-offset)))) ((looking-at "}") (- baseline rust-indent-offset)) ((and (nth 4 (syntax-ppss)) (looking-at "*")) (+ 1 baseline)) (t (or (if (> level 0) (progn (save-excursion ... ... ...))) (progn (back-to-indentation) (if (or ... ...) baseline (+ baseline rust-indent-offset))))))))
  (let ((indent (save-excursion (back-to-indentation) (let* ((level (rust-paren-level)) (baseline (if ... 0 ...))) (cond ((nth 3 ...) (let* ... ...)) ((looking-at "->") (save-excursion ... ...)) ((looking-at "}") (- baseline rust-indent-offset)) ((and ... ...) (+ 1 baseline)) (t (or ... ...))))))) (if indent (progn (if (<= (current-column) (current-indentation)) (indent-line-to indent) (save-excursion (indent-line-to indent))))))
  rust-mode-indent-line()
  indent-for-tab-command(nil)
  call-interactively(indent-for-tab-command nil nil)
  command-execute(indent-for-tab-command)
```

**Comment by mgalgs** [comment]
Another test case for what appears to be the same issue (though it has a slightly different backtrace):

```rust
#[test]
fn it_works() {
    let msg = "world";
    let msg = format!("hello,
{} //", msg);
    println!("{}", msg);
}
```

If I delete either the `//` or the `{}` I don't see the unbalanced paren error.

Backtrace:

```
Debugger entered--Lisp error: (scan-error "Unbalanced parentheses" 149 1)
  scan-sexps(150 -1)
  forward-sexp(-1)
  backward-sexp(1)
  funcall-interactively(backward-sexp 1)
  call-interactively(backward-sexp nil nil)
  command-execute(backward-sexp)
```

Repro:

Download [`emacs_clean.sh`](https://github.com/mgalgs/.emacs.d/blob/master/emacs_clean.sh) and [`emacs-clean.el`](https://github.com/mgalgs/.emacs.d/blob/master/emacs-clean.el).  Run:

    $ cargo new test
    $ ./emacs_clean.sh

then:

    M-x package-install
    rust-mode
    C-x C-f test/src/lib.rs
    (enter the text above)
    (go to the final { character)
    C-M-b

**Comment by MicahChalmer** [comment]
It looks like just the syntax table, without any of the rest of rust-mode, is sufficient to reproduce this on both of these examples.  The bracket matching, string finding, and indenting code are not relevant--the examples above confuse the sexp parsing and anything that uses it.  The following is the smallest repro I found:

````rust
{
    "
{} //"
}
````

It causes the last `}` not to match to its mate.

**Comment by MicahChalmer** [comment]
...and just because misery loves company, the same bug seems to affect `cc-mode` in C or C++ mode.  They use the same settings in the syntax table, so that's not surprising...but it doesn't bode well for trying to fix it.  It may be best thought of as a bug in Emacs itself--or at least in what appears to be the intended way to set up the comment syntax used by C, C++ and Rust.

**Comment by MicahChalmer** [comment]
Well I'll be...

https://www.gnu.org/software/emacs/manual/html_node/emacs/Left-Margin-Paren.html

"That's not a bug, it's a feature!"

tl;dr: Emacs has built into its scanning an assumption that an open paren at the left margin is the start of a top level definition.  This assumption can be disabled by setting `open-paren-in-column-0-is-defun-start` to `nil` (though the manual gives some vague warnings about performance if you do that.)

I've confirmed that setting that variable to nil does in fact make the examples above work correctly in rust-mode.  So this is not a bug in rust-mode but rather a wacky (but documented) "feature" of emacs.

**Comment by tromey** [comment]
I think it makes sense to set `open-paren-in-column-0-is-defun-start` to nil in `rust-mode`, and then see if it has any bad effects.  My impression is that it's only really relevant for extremely large files and mostly exists for cc-mode.
